### PR TITLE
Allow user to open team's context menu on header

### PR DESCRIPTION
**Pull request by ShenshunYao**

### CODE REVIEW

File: src/components/internal/Header.component.tsx [hunk 1]
```diff
@@ -70,9 +82,47 @@ export class HeaderComponent extends Component<Props, State> {
...
+                                Show invitations
+                            </li>
+                            <li
+                                className={styles.ContextMenuItem}
+                                onClick={this.onViewTeamSettingsClick}
+                            >
+                                Settings
+                            </li>
+                        </ul>
+                    )}
```
Comment (duanxianbo) [inline src/components/internal/Header.component.tsx]: can we move those to a seperate component?

File: src/components/internal/Header.component.tsx [hunk 2]
```diff
@@ -218,4 +268,66 @@ export class HeaderComponent extends Component<Props, State> {
...
+    private closeTeamContextMenu() {
+        this.setState({
+            showTeamContextMenu: false,
+            teamContextMenuLeft: -1000,
+            teamContextMenuTop: -1000,
+        });
+    }
+
+    private onAddMembersClick = () => {
+        this.props.onAddTeamMembersClick?.call(null, this.selectedTeamId || -1);
```
Comment (duanxianbo) [inline src/components/internal/Header.component.tsx]: if there is no valid teamId,  I think we just donot call the function instead of pass -1

**Review by duanxianbo — DISMISSED** [review]

File: src/components/internal/Header.component.tsx [hunk 3]
```diff
@@ -53,6 +62,9 @@ export class HeaderComponent extends Component<Props, State> {
             taskFiltersData: {
                 selectedTaskOwners: new Set<string>([userAll]),
             },
+            showTeamContextMenu: false,
+            teamContextMenuLeft: -1000,
```
Comment (naturetime10) [inline src/components/internal/Header.component.tsx]:
```suggestion
            teamContextMenuLeft: 0,
```

File: src/components/internal/Header.component.tsx [hunk 4]
```diff
@@ -53,6 +62,9 @@ export class HeaderComponent extends Component<Props, State> {
...
+            teamContextMenuTop: -1000,
```
Comment (naturetime10) [inline src/components/internal/Header.component.tsx]:
```suggestion
            teamContextMenuTop: 0,
```

File: src/components/internal/Header.component.tsx (hunk 3 above)
Comment (naturetime10) [reply src/components/internal/Header.component.tsx]: If contextMenu is not showing yet, I think we can set both left and top to zero.

File: src/components/internal/Header.component.tsx [hunk 5]
```diff
@@ -70,9 +82,47 @@ export class HeaderComponent extends Component<Props, State> {
...
-                    <div className={styles.TeamName}>
+                    <div className={styles.TeamName}
+                         onContextMenu={this.onTeamContextMenu(this.props.currentTeam?.id)}>
                         {this.props.currentTeam?.name}
                     </div>
+                    {this.state.showTeamContextMenu && (
+                        <ul
+                            ref={this.teamContextMenuRef}
+                            className={styles.TeamContextMenu}
+                            tabIndex={-1}
```
Comment (naturetime10) [inline src/components/internal/Header.component.tsx]: Just a knowledge check. Do you understand why we are setting `-1` here?

File: src/components/internal/Header.component.tsx (hunk 1 above)
Comment (naturetime10) [reply src/components/internal/Header.component.tsx]: I agree with @duanxianbo that we should consider refactoring into a separate component to encapsulate logic and to encourage code reuse.

File: src/components/internal/Header.component.tsx (hunk 2 above)
Comment (naturetime10) [reply src/components/internal/Header.component.tsx]: I think we should skip this function call if there is no valid team ID.

File: src/components/internal/Header.component.tsx [hunk 6]
```diff
@@ -218,4 +268,66 @@ export class HeaderComponent extends Component<Props, State> {
...
+        teamId: number | undefined,
+    ): (event: React.MouseEvent<HTMLElement>) => void => {
+        return (event: MouseEvent<HTMLElement>) => {
+            event.preventDefault();
+            this.showTeamContextMenu(teamId, event.clientX, event.clientY);
+        };
+    };
+
+    private showTeamContextMenu(teamId: number | undefined, left: number, top: number) {
+        this.selectedTeamId = teamId;
```
Comment (naturetime10) [inline src/components/internal/Header.component.tsx]: I think this not needed. you can directly read from `this.props.currentTeam?.id` in `onAddMembersClick`

File: src/components/internal/Header.component.tsx [hunk 7]
```diff
@@ -218,4 +268,66 @@ export class HeaderComponent extends Component<Props, State> {
...
+    }
+
+    private onTeamContextMenuBlur = () => {
+        this.closeTeamContextMenu();
+    };
+
+    private closeTeamContextMenu() {
+        this.setState({
+            showTeamContextMenu: false,
+            teamContextMenuLeft: -1000,
```
Comment (naturetime10) [inline src/components/internal/Header.component.tsx]:
```suggestion
            teamContextMenuLeft: 0,
```

File: src/components/internal/Header.component.tsx [hunk 8]
```diff
@@ -218,4 +268,66 @@ export class HeaderComponent extends Component<Props, State> {
...
+            teamContextMenuTop: -1000,
```
Comment (naturetime10) [inline src/components/internal/Header.component.tsx]:
```suggestion
            teamContextMenuTop: 0,
```

File: src/components/internal/Header.component.tsx [hunk 9]
```diff
@@ -218,4 +268,66 @@ export class HeaderComponent extends Component<Props, State> {
...
+
+    private onAddMembersClick = () => {
+        this.props.onAddTeamMembersClick?.call(null, this.selectedTeamId || -1);
+        this.closeTeamContextMenu();
+    };
+
+    private onShowMembersClick = () => {
+        this.props.onShowTeamMembersClick?.call(
+            null,
+            this.selectedTeamId || -1,
```
Comment (naturetime10) [inline src/components/internal/Header.component.tsx]: Skip this function call if `this.selectedTeamId` is undefined.

File: src/components/internal/Header.component.tsx [hunk 10]
```diff
@@ -218,4 +268,66 @@ export class HeaderComponent extends Component<Props, State> {
...
+        );
+        this.closeTeamContextMenu();
+    };
+
+    private onShowInvitationsClick = () => {
+        this.props.onShowInvitationsClick?.call(
```
Comment (naturetime10) [inline src/components/internal/Header.component.tsx]: Skip this function call if `this.selectedTeamId` is undefined.

File: src/components/internal/Header.component.tsx [hunk 11]
```diff
@@ -218,4 +268,66 @@ export class HeaderComponent extends Component<Props, State> {
...
+            this.selectedTeamId || -1,
+        );
+        this.closeTeamContextMenu();
+    };
+
+    private onViewTeamSettingsClick = () => {
+        this.props.onViewTeamSettingsClick?.call(
+            null,
+            this.selectedTeamId || -1,
+        );
```
Comment (naturetime10) [inline src/components/internal/Header.component.tsx]: Skip this function call if `this.selectedTeamId` is undefined.

**Review by naturetime10 — CHANGES_REQUESTED** [review]

File: src/components/internal/Header.component.tsx [hunk 12]
```diff
@@ -269,25 +234,61 @@ export class HeaderComponent extends Component<Props, State> {
...
-            this.showTeamContextMenu(teamId, event.clientX, event.clientY);
+            this.setState({
+                showTeamContextMenu: true,
+                teamContextMenuLeft: event.clientX,
+                teamContextMenuTop: event.clientY,
+            });
+            setTimeout(() => {
+                this.teamContextMenuRef.current?.focus();
+            });
+
```
Comment (naturetime10) [inline src/components/internal/Header.component.tsx]:
```suggestion
```

File: src/components/internal/Header.component.tsx [hunk 13]
```diff
@@ -297,37 +298,48 @@ export class HeaderComponent extends Component<Props, State> {
...
 
     private onAddMembersClick = () => {
-        this.props.onAddTeamMembersClick?.call(null, this.selectedTeamId || -1);
-        this.closeTeamContextMenu();
+        if (this.props.currentTeam?.id) {
+            this.props.onAddTeamMembersClick?.call(
+                null,
+                this.props.currentTeam.id,
+            );
+            this.closeTeamContextMenu();
```
Comment (naturetime10) [inline src/components/internal/Header.component.tsx]: We should always close context menu no matter current team ID is `undefined` or not.

File: src/components/internal/Header.component.tsx [hunk 14]
```diff
@@ -297,37 +298,48 @@ export class HeaderComponent extends Component<Props, State> {
...
-            null,
-            this.selectedTeamId || -1,
-        );
-        this.closeTeamContextMenu();
+        if (this.props.currentTeam?.id) {
+            this.props.onShowTeamMembersClick?.call(
+                null,
+                this.props.currentTeam.id,
+            );
+            this.closeTeamContextMenu();
```
Comment (naturetime10) [inline src/components/internal/Header.component.tsx]: We should always close context menu no matter current team ID is `undefined` or not.

File: src/components/internal/Header.component.tsx [hunk 15]
```diff
@@ -297,37 +298,48 @@ export class HeaderComponent extends Component<Props, State> {
...
-            null,
-            this.selectedTeamId || -1,
-        );
-        this.closeTeamContextMenu();
+        if (this.props.currentTeam?.id) {
+            this.props.onShowInvitationsClick?.call(
+                null,
+                this.props.currentTeam.id,
+            );
+            this.closeTeamContextMenu();
```
Comment (naturetime10) [inline src/components/internal/Header.component.tsx]: We should always close context menu no matter current team ID is `undefined` or not.

File: src/components/internal/Header.component.tsx [hunk 16]
```diff
@@ -297,37 +298,48 @@ export class HeaderComponent extends Component<Props, State> {
...
-            null,
-            this.selectedTeamId || -1,
-        );
-        this.closeTeamContextMenu();
+        if (this.props.currentTeam?.id) {
+            this.props.onViewTeamSettingsClick?.call(
+                null,
+                this.props.currentTeam.id,
+            );
+            this.closeTeamContextMenu();
```
Comment (naturetime10) [inline src/components/internal/Header.component.tsx]: We should always close context menu no matter current team ID is `undefined` or not.

**Review by naturetime10 — CHANGES_REQUESTED** [review]
I think it's okey to extract team context menu into a separate component in the next PR.

File: src/components/internal/Drawer.tsx [hunk 17]
```diff
@@ -35,8 +36,8 @@ export class DrawerUI extends Component<Props, State> {
         super(props);
         this.state = {
             showTeamContextMenu: false,
-            teamContextMenuLeft: -1000,
-            teamContextMenuTop: -1000,
+            teamContextMenuLeft: 0,
+            teamContextMenuTop: 0,
```
Comment (naturetime10) [inline src/components/internal/Drawer.tsx]: I think you don't need any of them. Let's remove those.

File: src/components/internal/Header.component.module.scss [hunk 18]
```diff
@@ -1,7 +1,9 @@
 @import '../../libs/ui/styleguide/easing';
+@import '../../libs/ui/styleguide/elevation';
 @import 'shared';
 
 @include header;
+$contextMenuSidePadding: 20px;
```
Comment (naturetime10) [inline src/components/internal/Header.component.module.scss]:
```suggestion
```

File: src/components/internal/TeamContextMenu.component.tsx [hunk 19]
```diff
@@ -0,0 +1,114 @@
...
+}
+
+interface State {
+    show: boolean;
+    top: number;
+    left: number;
+}
+
+export class TeamContextMenuComponent extends Component<Props, State> {
+    private contextMenuRef: RefObject<any> = createRef();
```
Comment (naturetime10) [inline src/components/internal/TeamContextMenu.component.tsx]:
```suggestion
    private contextMenuRef = createRef<HTMLULElement>();
```

**Review by naturetime10 — CHANGES_REQUESTED** [review]

File: src/components/internal/Drawer.tsx [hunk 20]
```diff
@@ -161,6 +161,7 @@ export class DrawerUI extends Component<Props, State> {
     };
 
     private onViewTeamSettingsClick = () => {
+        debugger;
```
Comment (naturetime10) [inline src/components/internal/Drawer.tsx]:
```suggestion
```

**Review by naturetime10 — CHANGES_REQUESTED** [review]

**Review by naturetime10 — APPROVED** [review]

File: src/components/internal/Drawer.tsx (hunk 17 above)
Comment (naturetime10) [inline src/components/internal/Drawer.tsx]: Those need to be removed.

**Review by naturetime10 — CHANGES_REQUESTED** [review]

**Review by naturetime10 — APPROVED** [review]

**Pull request merged**